codetransformer/tests/test_macros.py

```diff
@@ -0,0 +1,264 @@
+import pytest
+
+from functools import total_ordering
+from inspect import signature
+
+from .. import instructions as instrs
+from ..assembler import assemble_function
+from ..macros import AssertFail, IfStatement, ForLoop
+from ..utils.instance import instance
+
+
+def assert_same_result(f1, f2, *args, **kwargs):
+    try:
+        result1 = f1(*args, **kwargs)
+        f1_raised = False
+    except Exception as e:
+        result1 = e
+        f1_raised = True
+
+    try:
+        result2 = f2(*args, **kwargs)
+        f2_raised = False
+    except Exception as e:
+        result2 = e
+        f2_raised = True
+
+    if f1_raised and not f2_raised:
+        raise AssertionError("\n{} raised {}\n{} returned {}".format(
+            f1.__name__, result1, f2.__name__, result2
+        ))
+    elif not f1_raised and f2_raised:
+        raise AssertionError("\n{} returned {}\n {} raised {}".format(
+            f1.__name__, result1, f2.__name__, result2
+        ))
+    elif f1_raised and f2_raised:
+        assert type(result1) == type(result2) and result1.args == result2.args
+    else:
+        assert result1 == result2
+
+
+def test_simple_if_statement():
+
+    def goal(x, y):
+        z = 0
+        if x > y:
+            z = 1
+        z = z + 1
+        return z
+
+    assembly = [
+        instrs.LOAD_CONST(0),
+        instrs.STORE_FAST('z'),
+        IfStatement(
+            test=[
+                instrs.LOAD_FAST('x'),
+                instrs.LOAD_FAST('y'),
+                instrs.COMPARE_OP.GT,
+            ],
+            body=[
+                instrs.LOAD_CONST(1),
+                instrs.STORE_FAST('z'),
+            ],
+        ),
+        instrs.LOAD_FAST('z'),
+        instrs.LOAD_CONST(1),
+        instrs.BINARY_ADD(),
+        instrs.STORE_FAST('z'),
+        instrs.LOAD_FAST('z'),
+        instrs.RETURN_VALUE(),
+    ]
+    func = assemble_function(signature(goal), assembly)
+
+    assert_same_result(func, goal, 1, 1)
+    assert_same_result(func, goal, 1, 2)
+    assert_same_result(func, goal, 2, 1)
+    assert_same_result(func, goal, incomparable, 1)
+    assert_same_result(func, goal, 1, incomparable)
+
+
+def test_if_else():
+
+    def goal(x):
+        if x > 0:
+            return x
+        else:
+            return -x
+
+    assembly = [
+        IfStatement(
+            test=[
+                instrs.LOAD_FAST('x'),
+                instrs.LOAD_CONST(0),
+                instrs.COMPARE_OP.GT,
+            ],
+            body=[
+                instrs.LOAD_FAST('x'),
+                instrs.RETURN_VALUE(),
+            ],
+            else_body=[
+                instrs.LOAD_FAST('x'),
+                instrs.UNARY_NEGATIVE(),
+                instrs.RETURN_VALUE(),
+            ],
+        ),
+        AssertFail("Shouldn't ever get here!"),
```

Comment:
This is here because we currently can't compile a function that ends with just an `IfStatement`, because we emit a `JUMP_FORWARD` to a label after the else block.

One way to fix this might be to replace jumps to a final label with `LOAD_CONST(None); RETURN_VALUE()`.

Reply:
Though even that would be wasteful in this case, since the JUMP_FORWARD instruction is after an unconditional return inside the if block.  The thing we'd really want for this is a dead code elimination pass.

Reply:
We used to have an optimizer but it generated invalid code sometimes. We could always bring it back as an optional codetransformer.